classes/task/email_certificate_task.php

```diff
@@ -49,27 +49,81 @@ public function get_name() {
+        $lastprocessed = $taskprogress->last_processed;
+
         // Get all the certificates that have requested someone get emailed.
         $emailotherslengthsql = $DB->sql_length('c.emailothers');
         $sql = "SELECT c.*, ct.id as templateid, ct.name as templatename, ct.contextid, co.id as courseid,
-                       co.fullname as coursefullname, co.shortname as courseshortname
-                  FROM {customcert} c
-                  JOIN {customcert_templates} ct
+                    co.fullname as coursefullname, co.shortname as courseshortname
+                FROM {customcert} c
+                JOIN {customcert_templates} ct
                     ON c.templateid = ct.id
-                  JOIN {course} co
-                    ON c.course = co.id
-                 WHERE (c.emailstudents = :emailstudents
-                        OR c.emailteachers = :emailteachers
-                        OR $emailotherslengthsql >= 3)";
-        if (!$customcerts = $DB->get_records_sql($sql, ['emailstudents' => 1, 'emailteachers' => 1])) {
+                JOIN {course} co
+                    ON c.course = co.id";
+
+        // Add JOIN with mdl_course_categories to exclude certificates from hidden courses.
+        $sql .= " JOIN {course_categories} cat ON co.category = cat.id";
+
+        // Add conditions to exclude certificates from hidden courses.
+        $sql .= " WHERE (c.emailstudents = :emailstudents
+                 OR c.emailteachers = :emailteachers
+                 OR $emailotherslengthsql >= 3)";
+
+        // Check the includeinnotvisiblecourses configuration.
+        if (!$includeinnotvisiblecourses) {
+            // Exclude certificates from hidden courses.
+            $sql .= " AND co.visible = 1 AND cat.visible = 1";
+        }
+
+        // Add condition based on certificate execution period.
+        if ($certificateexecutionperiod > 0) {
+            // Include courses with no end date or end date greater than the specified period.
+            $sql .= " AND (co.enddate = 0 OR co.enddate > :enddate)";
+            $params['enddate'] = time() - $certificateexecutionperiod;
+        }
+
+        // Execute the SQL query.
+        if (!$customcerts = $DB->get_records_sql($sql, ['emailstudents' => 1, 'emailteachers' => 1] + $params)) {
             return;
         }
 
         // The renderers used for sending emails.
         $page = new \moodle_page();
         $htmlrenderer = $page->get_renderer('mod_customcert', 'email', 'htmlemail');
         $textrenderer = $page->get_renderer('mod_customcert', 'email', 'textemail');
-        foreach ($customcerts as $customcert) {
+
+        // Store the total count of certificates in the database.
+        $totalcertificatestoprocess = count($customcerts);
+        $DB->set_field('customcert_task_progress', 'total_certificate_to_process', $totalcertificatestoprocess, [
```

Comment:
The field is set here but it doesn't actually do anything. It just gets set then forgotten about afaict.

Reply:
This Field is used when Custom certificate settings > Scheduled task configuration > Certificates per run is set to any value other than Zero.